Complement phyint whitelist with blacklist

bket opened the pull request:
Fixes: #54

Implement new phyint configuration option (blacklist), which enables blocking of specific traffic.

pali commented on doc/igmpproxy.conf.5.in:
```diff
@@ -161,6 +161,21 @@ You may also specify whitelist entries for the upstream interface. Only igmp mem
+except that if a blacklist entry is defined, all igmp membership reports for
+that multicast group will be ignored.
+
+Note that if whitelist and blacklist overlap, blacklist will take precedence.
```
This looks to be strange behavior. Specially if you have big blacklist and need to whitelist some particular address from it.

Why not rather to take standard policy that the last matched rule take effect? This can be more useful for different kind of configurations.

This needs to answer two additions questions:
1. What happens when no whilelist / blacklist is specified? Everything is allowed or blocked? Allowed seems to be reasonable choice as it is backward compatible.
2. What is the default policy when particular address is not specified at all? Currently it is to block when whitelist keyword is used... Now I come up with idea that default policy may be blocked if all specified policy keywords are only whitelist (and at least one is specified), otherwise default is allowed ... this is backward compatible but I'm not sure how "sane" it is.

bket replied:
> This looks to be strange behavior. Specially if you have big blacklist and need to whitelist some particular address from it.

As it is, `whitelist` is currently doing two things: 1.) it ignores all igmp membership reports, 2.) except those whitelisted. `blacklist` does only one thing: it ignores blacklisted igmp membership reports. I chose to not only keep the behaviour of whitelist, but also make it independent from blacklist. Behaviour of whitelist can be changed, but we would risk breaking existing setups.

> Why not rather to take standard policy that the last matched rule take effect? This can be more useful for different kind of configurations.

If I'm not mistaken this would mean that the behaviour of `whitelist` changes depending on if there is a `blacklist` rule. See above.
Additionally, order of the rules is lost as whitelist and blacklist are stored in separate linked lists. For the 'last match rule' to work I need to either add an additional struct member (rule numer) to each list or put all rules in a single linked list and mark them (allow/block), and write additional checks. This would make the code more complex/fragile. My proposal is to maintaing the old behaviour of whitelist, and use blacklist to additionally ignore igmp membership reports. What do you think of the following change.

```suggestion
blacklist can be used to ignore additional igmp membership reports.
```

> This needs to answer two additions questions:

Guess it makes sense to add the following to the manpage.

```
Overview on the use of whitelist and blacklist:

whitelist    blacklist     igmp membership reports
not used     not used      allow all
not used     used          allow all except those blacklisted
used         not used      block all except those whitelisted
used         used          block all except those whitelisted, and not blacklisted
```

pali replied:
> If I'm not mistaken this would mean that the behaviour of whitelist changes depending on if there is a blacklist rule. See above.

Yes.

> Additionally, order of the rules is lost as whitelist and blacklist are stored in separate linked lists. For the 'last match rule' to work I need to either add an additional struct member (rule numer) to each list or put all rules in a single linked list and mark them (allow/block), and write additional checks.

Seems that change could be easy. Just extend `struct SubnetList` with a new member which says that it is blacklist or whitelist. `struct SubnetList` is used only for filtering and currently only whitelist operation.

And then it is just needed to traverse whole link list once and to the end and use rule which matches last time.

bket replied:
Thanks for your feedback. Really helpful! I extended `struct SubnetList` and changed to last-rule-wins.

Changes have been pushed.

pali commented on igmpproxy.conf:
```diff
@@ -33,9 +33,11 @@ phyint eth0 upstream  ratelimit 0  threshold 1
+## blocking SSDP / UPnP traffic
 ##------------------------------------------------------
 phyint eth1 downstream  ratelimit 0  threshold 1
+        blacklist 239.255.255.250/32
```
I do not like an idea to change default config file. I think that by default everything should be allowed and it is up to the admin to configure rounting as it is required per needs.

bket replied:
I will revert this bit after discussion on the bahaviour of whitelist.

bket replied:
Changes to the example config file have been reverted, and pushed.

bket committed:
Complement phyint whitelist with blacklist

Fixes: #54

Implement new phyint configuration option (blacklist), which enables
blocking of specific traffic.

pali commented on src/igmpproxy.h:
```diff
@@ -144,6 +144,7 @@ void my_log( int Serverity, int Errno, const char *FmtSt, ... );
     uint32_t            subnet_mask;
+    unsigned short      allow;
     struct SubnetList   *next;
 };
```
Put `allow` at the end of structure. On 64-bit systems, compiler has to align `next` pointer to be on 64-bit offset from beginning of structure, therefore it will change type of `allow` member from 16-bit (short) to 64-bit integer type.

Also igmpproxy already requires C99, so you can use `bool` type without any issue.

bket replied:
Forgot about allignment. Thanks!

`allow` has been moved and changed to `bool`,

pali commented on src/config.c:
```diff
@@ -355,6 +355,22 @@ struct vifconfig *parsePhyintToken(void) {
+            *agrpPtr = parseSubnetAddress(token);
+            if(*agrpPtr == NULL) {
+                parseError = 1;
+                my_log(LOG_WARNING, 0, "Unable to parse subnet address.");
```
This should be (fatal) error. When config line is invalid igmpproxy should not try to ignore or change keyword to something else. (Same applies for `whitelist` keyword parsing).

bket replied:
`blacklist` and `whitelist` now give an error when parsing fails.

pali commented on src/request.c:
```diff
@@ -83,22 +83,27 @@ void acceptGroupReport(uint32_t src, uint32_t group) {
+
+        if(match == NULL || (match != NULL && match->allow == 1)) {
+            // The membership report was OK... Insert it into the route table..
+            insertRoute(group, sourceVif->index, src);
```
This has an issue. If config file contains only `whitelist` keywords then everything is allowed, even those ranges which are not listed in `whitelist` keyword.

bket replied:
Yes, again you are correct. Guess I did not test enough. Issue seems solved now.

pali commented on src/request.c:
```diff
@@ -83,22 +83,37 @@ void acceptGroupReport(uint32_t src, uint32_t group) {
-        struct SubnetList *sn;
-        for(sn = sourceVif->allowedgroups; sn != NULL; sn = sn->next)
+        bool                 allow_list = 0;
+        bool                 block_list = 0;
```
C99 supports boolean keyword`false`.

bket replied:
Changed to keyword...

pali commented on src/request.c:
```diff
@@ -83,22 +83,37 @@ void acceptGroupReport(uint32_t src, uint32_t group) {
+
+        if((!allow_list && block_list &&
+          (match == NULL || (match != NULL && match->allow ))) ||
+          (allow_list && match != NULL && match->allow)) {
```
This condition does not make sense... Condition `(match != NULL && match->allow )` is never true because at the beginning you check that `!allow_list` which is true only when `match->allow` is false.

bket replied:
I'm clearly not seeing what you are seeing. Your guidiance is much appreciated.

After rewriting I come to the following
```
if((!allow_list && match == NULL) ||
  (allow_list && match != NULL && match->allow))
```

Does this makes sense?

pali replied:
Yes, this logic looks to be usable. Just do not forget to document how whitelist/blacklist is evaluated.

bket replied:
Evaluation has been documented in the manpage.

pali commented on src/rttable.c:
```diff
@@ -157,18 +157,24 @@ static void sendJoinLeaveUpstream(struct RouteTable* route, int join) {
-                if (sn == NULL) {
+                if((block_list && match != NULL && !match->allow) ||
+                  (!block_list && match == NULL)) {
                     my_log(LOG_INFO, 0, "The group address %s may not be forwarded upstream. Ignoring.", inetFmt(group, s1));
```
This condition evaluated incorrectly for address `3` and following configuration:
```
whitelist 1
blacklist 2
```
Because `block_list` is set to true and `match` to `NULL`. Which results in forwarding address `3`. But based on logic in `request.c` and description in manpage it should not be forwarded.

bket replied:
Damnit, I'm starting to regret this PR ;-)

Should be fixed now.

I copied the evaluation from request.c, and negate the result.

```
                // Keep in sync with request.c, note the negation
                if(!((!allow_list && match == NULL) ||
                  (allow_list && match != NULL && match->allow))) {
```

pali replied:
Ok, this seems to be correct now.

And, well this is probably reason why nobody wanted to fix issue #54 for a longer time...

pali commented:
Thanks, merged!
